### ISSUE TITLE
More Augmentations

### ISSUE THREAD

**Issue by jaideep11061982** [ultralytics/yolov5#3699]
## ❔Question
How i can use more and more augmentations related to Image Enhancements.

**Comment by glenn-jocher** [comment]
@jaideep11061982 what augmentations are you interested in?

**Comment by tetelevm** [comment]
@glenn-jocher
I am also interested in this question. As I understand it, YOLO slices images for training, but does it also know how to rotate images by `(90, 180, 270)` degrees? (And is it even worth doing that?)

**Comment by glenn-jocher** [comment]
@tetelevm hyperparameters for augmentation including rotation are defined in the hyp file:
https://github.com/ultralytics/yolov5/blob/fbf41e09134b113f8e79ae01b4eee40d00797b2d/data/hyp.scratch.yaml#L22-L33

**Comment by decent-engineer-decent-datascientist** [comment]
@glenn-jocher have you checked this out:
https://github.com/facebookresearch/AugLy

I'd never need this for my use cases, but you can do stuff like add ig filters, emojis, simulate screenshots, etc.
A lot of modern digital-life augmentations that might not win you any AP on COCO, but would be useful for real-world applications.

Here's a list of augmentations:
https://github.com/facebookresearch/AugLy/blob/main/augly/image/__init__.py

**Comment by glenn-jocher** [comment]
@decent-engineer-decent-datascientist yes I've seen the facebook augmentation package, which seems to be the flavor of the week, but I would recommend Albumentations:
https://github.com/albumentations-team/albumentations

You could integrate additional Albumentations augmentations in the YOLOv5 dataloader here:
https://github.com/ultralytics/yolov5/blob/e8810a53e83ddb5dd6bf8e871c2ede701007047c/utils/datasets.py#L536-L552

**Comment by glenn-jocher** [comment]
the best place to insert Albumentation augmentations into the YOLOv5 dataloader is here:
https://github.com/ultralytics/yolov5/blob/63060910a68bfde238872d629ab88e2e7bc736e8/utils/datasets.py#L536-L552

where `img` is the image and `labels` are the bounding box labels. Note that any Albumentation augmentations you add will be in addition to the existing automatic YOLOv5 augmentations defined in your hyperparameter file:
https://github.com/ultralytics/yolov5/blob/63060910a68bfde238872d629ab88e2e7bc736e8/data/hyp.scratch.yaml#L21-L33

**Comment by glenn-jocher** [comment]
@decent-engineer-decent-datascientist see PR #3882 for a proposed automatic Albumentations integration.

**Comment by glenn-jocher** [comment]
@decent-engineer-decent-datascientist @jaideep11061982 @tetelevm good news 😃! Your original issue may now be fixed ✅ in PR #3882. This PR implements a [YOLOv5](https://github.com/ultralytics/yolov5) 🚀 + [Albumentations](https://github.com/albumentations-team/albumentations) integration. The integration will automatically apply Albumentations transforms during YOLOv5 training if `albumentations>=1.0.0` is installed in your environment.

## Get Started

To use albumentations simply `pip install -U albumentations` and then update the augmentation pipeline as you see fit in the `Albumentations` class in `yolov5/utils/augmentations.py`. Note these Albumentations operations run **in addition to** the YOLOv5 **hyperparameter** augmentations, i.e. defined in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch.yaml).

```python
class Albumentations:
    # YOLOv5 Albumentations class (optional, used if package is installed)
    def __init__(self):
        self.transform = None
        try:
            import albumentations as A
            check_version(A.__version__, '1.0.0')  # version requirement

            self.transform = A.Compose([
                A.Blur(p=0.1),
                A.MedianBlur(p=0.1),
                A.ToGray(p=0.01)],
                bbox_params=A.BboxParams(format='yolo', label_fields=['class_labels']))

            logging.info(colorstr('albumentations: ') + ', '.join(f'{x}' for x in self.transform.transforms))
        except ImportError:  # package not installed, skip
            pass
        except Exception as e:
            logging.info(colorstr('albumentations: ') + f'{e}')

    def __call__(self, im, labels, p=1.0):
        if self.transform and random.random() < p:
            new = self.transform(image=im, bboxes=labels[:, 1:], class_labels=labels[:, 0])  # transformed
            im, labels = new['image'], np.array([[c, *b] for c, b in zip(new['class_labels'], new['bboxes'])])
        return im, labels
```

## Example Result

Example `train_batch0.jpg` on COCO128 dataset with Blur, MedianBlur and ToGray. See the YOLOv5 [Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) to reproduce: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

![train_batch0](https://user-images.githubusercontent.com/26833433/124400879-ff331b80-dd25-11eb-9b67-fe85ac4ca104.jpg)


# Update

To receive this YOLOv5 update:

- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
